### ISSUE TITLE
Lab Blockbattle: event handling is not cross-platform compatible

### ISSUE THREAD

**Issue by emmanuelkring** [lunduniversity/introprog#491]
The handling of events in lab blockbattle is not cross-platform compatible. The problem is that the string that a `BlockWindow.Event.KeyPressed` object is created with differs on different operating systems, at least for the arrow keys. For example, on Linux pressing the up arrow creates a `KeyPressed` object with the string "Up" while on MacOS the string will be "↑". 

One possible fix is to simply mention this in exercise 7 e) in chapter 5 and let the students copy the string that appears in the terminal. The downside is that the code will not be cross-platform compatible. 

Another possible fix is to make changes to pixelwindow so that this problem does not occur. This might require some changes to the lab as well but the code will then be cross-platform compatible. 

Other labs might be affected by this issue too.

**Comment by jos-rosenqvist-3812** [comment]
https://github.com/lunduniversity/introprog-scalalib/issues/1 is related.

**Comment by bjornregnell** [comment]
We need someone with macos who can give a list of utf codes and provide a table with codes like \u2190 for left arrow or whatever. Perhaps you can fix that @emmanuelkring ? PR welcome. @jos-rosenqvist-3812 perhaps you can assist @emmanuelkring with review etc if  @emmanuelkring  can take this on?

**Comment by bjornregnell** [comment]
AND before we do some special thing for Apple we should investigate if it is possible to hack PixelWindow to detect if on macos and fix clear text alternatives exposed to PixelWindow users, see also https://github.com/lunduniversity/introprog-scalalib/issues/1

**Comment by bjornregnell** [comment]
Work-around until fixed: use keys: JKLI instead of arrows

**Comment by Patrikgyllvin** [comment]
Another issue with this is that the string representation is dependent on the locale of the JVM as well on Linux, which means `java.awt.event.KeyEvent.getKeyText()` might return for the up arrow key e.g "Up" (en_US) or "Upp" (sv_SE). The issue on macOS is also related to locale.

I have what I believe is a working fix for both issues which uses lookup for the special keys for every OS using the awt virtual key codes `java.awt.event.KeyEvent.VK_*`. This would also solve lunduniversity/introprog-scalalib#1. I have tested it working on macOS, will test on Linux as well.

Other possible solutions could include using these virtual key codes directly with PixelWindow instead, i.e keeping an attribute `var _lastKey: Int`, or trying to set the locale to e.g en_US and not using unicode. However setting the locale is VM-global.

**Comment by bjornregnell** [comment]
I'm not sure what's best; e.g. working with dates is also depending on locale so setting this to _US might have other consequences. @Patrikgyllvin let me know when you have tested this on Linux and we can discuss what's best before you make a PR. Ok?

**Comment by Patrikgyllvin** [comment]
@bjornregnell I have tested it on Linux and it is working. This map lookup approach might be better as the string representations become uniform across all OSs and locales. We would need to decide on a name for the Cmd/Windows/Super key as it produces a `VK_META` key code on macOS and `VK_WINDOWS` on Linux/Windows.

There is an issue where `TAB` keystrokes are not recognized by introprog, which seems related to the Focus Subsystem of Swing. Maybe open another issue?

**Comment by bjornregnell** [comment]
@Patrikgyllvin Yes open a new issue on introprog-scala lib and mention the existing Delete-issue and this issue and we continue discuss there. My suggestion is that Cmd/Window/Super-key shoud be called just "Meta" without VK. This will end up as a PR on introprog-scalalib...

**Comment by Patrikgyllvin** [comment]
@bjornregnell That seems like a good idea, and is how I have them mapped right now.

Opened new issue lunduniversity/introprog-scalalib#11.
